pandaproxy/sr: add simple benchmarks for sr api

IoannisRP opened the pull request:
Implements:
- [CORE-9857](https://redpandadata.atlassian.net/browse/CORE-9857)
- [CORE-9858](https://redpandadata.atlassian.net/browse/CORE-9858)

Add initial benchmarks for the schema registry api.
This is meant to be extended in time with the rest of the api endpoints.

In this first pass, only `POST /subjects/{}` and `POST /subjects/{}/versions` with default parameters are called.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none


[CORE-9857]: https://redpandadata.atlassian.net/browse/CORE-9857?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CORE-9858]: https://redpandadata.atlassian.net/browse/CORE-9858?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
pandaproxy/sr: add simple benchmarks for sr api

IoannisRP commented on src/v/pandaproxy/schema_registry/test/api_bench.cc:
```diff
@@ -0,0 +1,177 @@
+    return pps::subject{ss::format("TestSubject{}", i_sub)};
+}
+
+void setup_client(::http::client& client, int n_subjects, int n_versions) {
+    for (int i_sub = 0; i_sub < n_subjects; ++i_sub) {
+        pps::subject sub = make_subject(i_sub);
+        for (int i_ver = 1; i_ver <= n_versions; ++i_ver) {
+            auto schema = make_proto_schema(sub, pps::schema_version{i_ver});
+            auto payload = make_payload(schema);
+            auto res = post_schema(client, sub, payload);
+            BOOST_REQUIRE_EQUAL(
+              res.headers.result(), boost::beast::http::status::ok);
+        }
+    }
+}
```
Most of the time of the benchmarks are spent setting up sr. This could be replaces with producing directly to `_schemas` in the future, to allow for bigger test cases, but it should be good enough as a first pass.

IoannisRP commented on src/v/pandaproxy/schema_registry/test/BUILD:
```diff
@@ -374,3 +374,19 @@ redpanda_cc_gtest(
     ],
 )
+
+redpanda_cc_bench(
+    name = "api_rpbench",
+    srcs = [
+        "api_bench.cc",
+    ],
+    deps = [
+        ":client_utils",
+        "//src/v/http",
+        "//src/v/json",
+        "//src/v/pandaproxy",
+        "//src/v/pandaproxy/test:fixture",
+        "@abseil-cpp//absl/strings",
+        "@seastar//:benchmark",
+    ],
+)
```
When running this test with multiple cores, i saw lot's of noise which can be attributed to inter-shard communications. Opted for 1 cpu (which is the default in bazel) to start with measuring the cost of the algorithms without worrying about inter-shard costs, at this point.

StephanDollberg replied:
Going with just 1 cpu definitely the best way to start. There is also some measurement bugs in the test frame when using multiple cores.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64625-->
<!--6ab1370f-542a-40db-9af1-690d5414df56-->
<!--749a5763-bc56-4d56-8a7e-3a0aa01a1f2e-->
<!--51a47022-4cfa-4324-81a1-08c40fc2fce6-->
<!--1f6d97a8-de93-4991-81ae-f3b2c4b0c764-->
<!--title-->
## Retry command for [Build#64625](https://buildkite.com/redpanda/redpanda/builds/64625)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_decommissioning_working_node@{"delete_topic":false,"tick_interval":5000}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/datalake/throttling_test.py::DatalakeThrottlingTest.test_basic_throttling@{"catalog_type":"nessie","cloud_storage_type":1}
tests/rptest/tests/datalake/throttling_test.py::DatalakeThrottlingTest.test_basic_throttling@{"catalog_type":"rest_hadoop","cloud_storage_type":1}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--839d2095-588a-473f-bfe5-ba345f813585-->
<!--TEST RESULTS-->
<!--fd7b5854-915e-4ace-ba87-38451bfb2820-->
<!--TEST RESULTS-->
<!--e67269d1-e759-4f89-ad34-5be8353d3733-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64625">build#64625</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64625#01963fd5-9c7f-461b-b8a7-0a7ac15231e4 | FLAKY | 14/21 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/64625#01963ff1-4ec2-4352-9f44-4fe7794b6cf1 | FLAKY | 16/21 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/64625#01963ff1-4ec2-432f-9662-8f1304ac1a01 | FLAKY | 16/21 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=False.tick_interval=5000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64625#01963fd5-9c81-483f-8568-e296a6b63ee4 | FLAKY | 20/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64668">build#64668</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64668#01964355-2740-4208-8233-dcb19223c6f7 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64668#019643ab-f5d5-490f-a7f0-379d509693dc | FLAKY | 17/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64668#01964399-de15-4aa9-b3df-452e05eb5c68 | FLAKY | 16/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64668#01964399-de16-415b-a86d-7bb2ad80b6a4 | FLAKY | 17/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64751">build#64751</a></summary>


| test_id                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit        | https://buildkite.com/redpanda/redpanda/builds/64751#019648d3-2cd8-4334-aa6a-b9ae9e0fc26c | FLAKY         | 1/2      |
</details>
<!--/body-->

BenPope commented on src/v/pandaproxy/schema_registry/test/api_bench.cc:
```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
```
nitpick: year

BenPope commented on src/v/pandaproxy/schema_registry/test/BUILD:
```diff
@@ -374,3 +374,19 @@ redpanda_cc_gtest(
+
+redpanda_cc_bench(
+    name = "api_rpbench",
+    srcs = [
```
I needed a longer timeout.

IoannisRP replied:
i forgot that we have an `sh_test` running as part of the benchmark as well. Based on my current local timings, it should be `moderate` (instead of `short` which is the default), so i will change it to that.

BenPope commented on src/v/pandaproxy/schema_registry/test/api_bench.cc:
```diff
@@ -0,0 +1,177 @@
+
+namespace {
+
+ss::sstring make_proto_schema(const pps::subject& sub, pps::schema_version v) {
```
This doesn't really represent a `schema_version`, it's a field count

BenPope commented on src/v/pandaproxy/schema_registry/test/api_bench.cc:
```diff
@@ -0,0 +1,177 @@
+            auto schema = make_proto_schema(sub, pps::schema_version{i_ver});
+            auto payload = make_payload(schema);
+            auto res = post_schema(client, sub, payload);
+            BOOST_REQUIRE_EQUAL(
+              res.headers.result(), boost::beast::http::status::ok);
```
Does this do anything? I can't seem to be able to break it, even when I send an invalid request.

IoannisRP replied:
I copied it from another bench and didn't bother to test it :sweat: 
Yeah, it appears it does nothing. The check fails, but nothing comes out of it. I will remove it.

BenPope commented on src/v/pandaproxy/schema_registry/test/api_bench.cc:
```diff
@@ -0,0 +1,177 @@
+  const ss::sstring& schema,
+  pps::schema_type type = pps::schema_type::protobuf) {
+    return ss::format(
+      R"({{ "schemaType": "{}", "schema": "{}" }})",
+      type,
+      absl::CEscape(schema));
```
I wonder if it'd be sufficient to use:
```suggestion
      R"({{ "schemaType": "{}", "schema": "{:?} }})", type, schema);

IoannisRP replied:
it should be, but `24.2.x` doesn't support `:?` yet. This is not meant to be back-ported, just playing it safe.
Also, while they do the same, `:?` is quite poorly documented, so i am good to keep not using it till they fix that. I mainly dislike the misnomer in its name. It is called "debug" formatting (or something) while it does escaping.

BenPope reviewed (commented):
LGTM

There is some inconsistent use of `ss::format`, `fmt::format`.

StephanDollberg commented on src/v/pandaproxy/schema_registry/test/api_bench.cc:
```diff
@@ -0,0 +1,177 @@
+
+} // namespace
+
+class schema_registry_bench_fixture : public pandaproxy_test_fixture {
```
Just a nit really but because the full test name in CPD etc. will be  `pandaproxy_schema_registry_api_bench_schema_registry_bench_fixture_lookup_x1_1` you could make the fixture class name be something shorter. It's fine either way though.

IoannisRP replied:
i'll shorten it :+1:

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/7ed512e33365b97bf713f69c3806dd5c6532f0f4..f1dac00157903dd1f81560f90eea108ab640e26b):
- changes `fmt::format` to `ss::format`
- replaced `BOOST_REQUIRE_EQUAL` with `vassert`
- shorted name of fixture
- Fixed benchmarks. They should all be querying for the first subject, which is `0`, not `1`.
- Fixed timeout to `moderate`

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64668-->
<!--165a0dde-7656-4d7a-b32f-f110cbe6cbca-->
<!--title-->
## Retry command for [Build#64668](https://buildkite.com/redpanda/redpanda/builds/64668)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
```
<!--/body-->

michael-redpanda commented:
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
